### PR TITLE
[Python] Get available python version and use it for Python SDK harness boot entry point 

### LINKED ISSUE #26958: [Bug]: Beam's Python SDK harness boot entrypoint may fail when `python` is not accessible, but `python3` is.
### What happened?

The code in 

https://github.com/apache/beam/blob/98210d99b8530346b66fcffe66b893924c910bea/sdks/python/container/boot.go#L94

assumes the Python executable is always `python`, which seems to not work on Ubuntu system. We should be able to detect that and handle this case gracefully. 

There were also asks in the past to consider making  Python path configurable .  We could introduce an environment variable that controls the path to interpreter, defer to it if specified, if not - try python, if that fails, try python3.

### Issue Priority

Priority: 3 (minor)

### Issue Components

- [x] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by riteshghorse**
Use `GetPythonVersion` utility to find the available python command to use. If `PYTHON_PATH` is set then it uses that.

Fixes #26958 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by riteshghorse** [comment]
R: @tvalentyn

**Comment by riteshghorse** [comment]
Oh I think we would need to add a flag to configure python path. But this PR just adds the utility to read that env variable and choosing available python version otherwise

File: sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download_test.go [hunk 1]
```diff
@@ -217,3 +217,31 @@ func TestGetJar_dev(t *testing.T) {
...
+			PYTHON_PATH: "/bin/python",
+		},
+		{
+			name:        "PYTHON_PATH not set",
+			PYTHON_PATH: "",
+		},
+	}
+
+	for _, test := range tests {
+		os.Setenv("PYTHON_PATH", test.PYTHON_PATH)
```
Comment (tvalentyn) [inline sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download_test.go:237]: Consider setting only if not empty.

File: sdks/python/container/boot.go [hunk 2]
```diff
@@ -91,7 +92,12 @@ func main() {
 			"--container_executable=/opt/apache/beam/boot",
 		}
 		log.Printf("Starting worker pool %v: python %v", workerPoolId, strings.Join(args, " "))
-		if err := execx.Execute("python", args...); err != nil {
+		pythonVersion, err := expansionx.GetPythonVersion()
+		if err != nil {
+			log.Print("Python SDK worker pool exited.")
```
Comment (tvalentyn) [inline sdks/python/container/boot.go:97]: Can we buble up the "no python installation found" error instead?
Should we use  `log.Fatalf` instead?

File: sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go [hunk 3]
```diff
@@ -382,7 +382,14 @@ func jarExists(jarPath string) bool {
 	return err == nil
 }
 
-func getPythonVersion() (string, error) {
+// GetPythonVersion returns the Python version to use. It checks for
+// env variable PYTHON_PATH and returns that it if set.
+// If no PYTHON_PATH is defined then it checks for python or python3
```
Comment (tvalentyn) [inline sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go:387]: Let's modify the error message on line 399 to say sth like: 

no python installation found. If you use a custom container image, please verify that python or python3 command launches the interpreter  or specify the full path to the interpreter file in PYTHON_PATH environment variable.

File: sdks/python/container/boot.go [hunk 4]
```diff
@@ -336,7 +342,11 @@ func setupVenv(ctx context.Context, logger *tools.Logger, baseDir, workerId stri
 	if err := os.MkdirAll(dir, 0750); err != nil {
 		return "", fmt.Errorf("failed to create Python venv directory: %s", err)
 	}
-	if err := execx.Execute("python", "-m", "venv", "--system-site-packages", dir); err != nil {
+	pythonVersion, err := expansionx.GetPythonVersion()
+	if err != nil {
+		return "", fmt.Errorf("python installation not found: %v", err)
```
Comment (tvalentyn) [inline sdks/python/container/boot.go:347]: should we return `err` without modifications here this message is already included?

**Review by tvalentyn — CHANGES_REQUESTED** [review]
Thanks a lot!

**Comment by tvalentyn** [comment]
> Oh I think we would need to add a flag to configure python path. 

You mean a flag for boot.go? I think having env variable is good enough.

**Comment by riteshghorse** [comment]
> > Oh I think we would need to add a flag to configure python path.
> 
> You mean a flag for boot.go? I think having env variable is good enough.

Got it. I was wondering if the flag set on local machine would be picked up on the worker.

**Comment by tvalentyn** [comment]
no, it would have to be set in Dockerfile using `ENV` command.

File: sdks/python/container/boot.go (hunk 2 above)
Comment (riteshghorse) [reply sdks/python/container/boot.go:97]: yeah

File: sdks/python/container/boot.go (hunk 4 above)
Comment (riteshghorse) [reply sdks/python/container/boot.go:347]: done

File: sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download_test.go (hunk 1 above)
Comment (riteshghorse) [reply sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download_test.go:237]: done

**Comment by riteshghorse** [comment]
> > > Oh I think we would need to add a flag to configure python path.
> > 
> > 
> > You mean a flag for boot.go? I think having env variable is good enough.
> 
> Got it. I was wondering if the flag set on local machine would be picked up on the worker.

OK, I'll send a CL to update the custom container doc as well near 2.51.0 release

**Comment by riteshghorse** [comment]
Run PythonDocs PreCommit

**Comment by riteshghorse** [comment]
Failing tests are not related to this PR

File: sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go [hunk 5]
```diff
@@ -396,7 +396,9 @@ func GetPythonVersion() (string, error) {
 			return v, nil
 		}
 	}
-	return "", fmt.Errorf("no python installation found")
+	return "", errors.New("no python installation found. If you use a" +
```
Comment (tvalentyn) [inline sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go:399]: add spaces before closing quotes.

**Review by tvalentyn — APPROVED** [review]

**Comment by riteshghorse** [comment]
Run Python Dataflow ValidatesContainer

**Comment by riteshghorse** [comment]
Run Python_Runners PreCommit

**Comment by riteshghorse** [comment]
Without python inter

> 

If the python interpreter is not available then it will fail at the `pip install apache-beam[gcp]` step itself when building the container

**Comment by tvalentyn** [comment]
> If the python interpreter is not available then it will fail at the pip install apache-beam[gcp] step itself when building the container

you can build a regular container, then manually build another container:

```
FROM buster
COPY --from=<your_container_image> /opt/apache/beam /opt/apache/beam

ENTRYPOINT ["/opt/apache/beam/boot"]
```

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesContainer

**Comment by tvalentyn** [comment]
VC run passed before: https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont_PR/306/

**Comment by riteshghorse** [comment]
The error message we wrote is logged for this [job](https://pantheon.corp.google.com/dataflow/jobs/us-central1/2023-08-23_09_49_03-4656118819153030201;bottomTab=WORKER_LOGS;graphView=0;logsSeverity=INFO?pageState=(%22dfTime%22:(%22s%22:%222023-08-23T16:49:04.304Z%22,%22e%22:%222023-08-23T16:49:23.391Z%22))&project=google.com:clouddfe)

image

using the dockerfile:
```
FROM debian

COPY --from=us.gcr.io/google.com/clouddfe/riteshghorse/beam_python3.9_sdk:python_custom /opt/apache/beam /opt/apache/beam

ENTRYPOINT ["/opt/apache/beam/boot"]
```

**Comment by riteshghorse** [comment]
Run PythonDocs PreCommit

**Pull request merged**